torch/distributed/device_mesh.py

```diff
@@ -78,30 +80,68 @@ def get_current_mesh(self) -> "DeviceMesh":
             return self.mesh_stack[-1]
 
         def create_sub_mesh(
-            self, device_mesh: "DeviceMesh", submesh_dim_names: Tuple[str, ...]
+            self,
+            device_mesh: "DeviceMesh",
+            submesh_dim_names: Tuple[str, ...],
+            submesh_dims: List[Tuple[int, ...]],
         ) -> "DeviceMesh":
-            if device_mesh != self.get_root_mesh(device_mesh):
-                raise RuntimeError("Cannot create a submesh from a submesh.")
-
-            # submesh_dims are the mesh dimension of the submesh in the device mesh.
-            submesh_dims = [
-                not_none(device_mesh.mesh_dim_names).index(mesh_dim_name)
-                for mesh_dim_name in submesh_dim_names
-            ]
-            submesh_dim_sizes = [
-                device_mesh.mesh.size(mesh_dim) for mesh_dim in submesh_dims
+            # Get the submesh dim size from the submesh_dims.
+            # For example, if we have a 3D mesh with mesh_shape (2, 2, 2) mesh_dim_names ("dp", "cp", "tp") and we want
+            # to slice out mesh["dp_cp"], then submesh_dims = [(0, 1), (2,)] and submesh_dim_size = [2 * 2, 2] = [4, 2].
+            # If we want to slice out mesh["dp", "cp"], then submesh_dims = [(0,), (1,)] and submesh_dim_size = [2, 2].
+            slice_dim_size = [
+                reduce(
+                    lambda x, y: device_mesh.mesh.size(x) * device_mesh.mesh.size(y),
+                    mesh_dim,
+                )
+                if len(mesh_dim) > 1
+                else device_mesh.mesh.size(mesh_dim[0])
+                for mesh_dim in submesh_dims
             ]
 
-            mesh_dims_remained = list(range(device_mesh.mesh.ndim))
-            for submesh_dim in submesh_dims:
-                mesh_dims_remained.remove(submesh_dim)
+            mesh_tensor = device_mesh.mesh
+            # slice_dim_idx could be differnt from submesh_dims, as we may need to flatten out some dims.
```

Comment:
Maybe add one more comment to show that the indices in `slice_dim_idx` are the indices of `mesh_tensor`.